pkg/roachprod/vm/ibm/provider.go

```diff
@@ -0,0 +1,909 @@
+
+	// We will create all VMs in parallel. We don't use an errgroup here
+	// because we don't want goroutines to be cancelled if one of them fails.
+	// This is because we create the instance and tag it in two different
+	// API calls, and we want to make sure that all instances are tagged
+	// even if an error occurs during the creation of one instance.
+	var mutex syncutil.Mutex
+	var g sync.WaitGroup
+	createErrors := make([]error, 0)
+	vms := make(vm.List, len(names))
+
+	for i, vmName := range names {
+
+		zone := zones[i]
+
+		region, err := p.zoneToRegion(zone)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to compute zone from region")
+		}
+
+		// Get the image ID for the region.
+		imageID, err := p.getImageIDForRegion(l, providerOpts.ImageAMI, region)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to get image ID for region %s", region)
+		}
+
+		// Get the SSH key ID for the region.
+		sshKeyID, err := p.getSshKeyID(l, keyName, region)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to get SSH key ID for region %s", region)
+		}
+
+		g.Add(1)
+
+		go func() {
+			defer g.Done()
+
+			vm, err := p.createInstance(l, instanceOptions{
+				vmName: vmName,
+				region: region,
+				zone:   zone,
+				infraOpts: &infraOpts{
+					vpcID:    p.config.regions[region].vpcID,
+					subnetID: p.config.regions[region].zones[zone].subnetID,
+					imageID:  imageID,
+					sshKeyID: sshKeyID,
+				},
+				vmOpts:       opts,
+				providerOpts: providerOpts,
+				tags:         tags,
+			})
+
+			mutex.Lock()
+			defer mutex.Unlock()
+			if err != nil {
+				createErrors = append(
+					createErrors,
+					errors.Wrapf(
+						err,
+						"failed to create instance %s in region %s", vmName, region,
+					),
+				)
+			} else {
+				vms[i] = *vm
+			}
+		}()
+	}
+
+	g.Wait()
+
+	if len(createErrors) > 0 {
+		return nil, errors.Join(createErrors...)
+	}
+
+	return vms, nil
+}
+
+// Delete is part of the vm.Provider interface.
+// It deletes the VMs in the list.
+func (p *Provider) Delete(l *logger.Logger, vms vm.List) error {
+
+	// We don't delete floating IP addresses as they are billed per creation.
+	// We want to keep a pool of unbound floating IP addresses to use for
+	// future instances.
+	// If we want to delete floating IPs, we need to set the flag to true.
+	deleteFloatingIPs := false
+
+	var g errgroup.Group
+	for _, vm := range vms {
+
+		g.Go(func() error {
+
+			zone := vm.Zone
+
+			// We need to query the instance to get the floating IP address
+			// and the attached volumes.
+			instance, err := newInstanceFromCRN(p, vm.ProviderID, deleteFloatingIPs)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get instance from CRN %s", vm.ProviderID)
+			}
+
+			vpcService, err := p.getVpcServiceFromZone(zone)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get VPC service for zone %s", zone)
+			}
+
+			// Start by deleting the floating IP address if any
+			if deleteFloatingIPs {
```

Comment:
Nit: maybe factor this out into a helper especially since it's statically disabled.